File: crates/liteparse/src/parser.rs
use crate::config::{LiteParseConfig, parse_target_pages};
#[cfg(not(target_arch = "wasm32"))]
use crate::conversion;
use crate::error::LiteParseError;
use crate::extract;
use crate::ocr::OcrEngine;
#[cfg(not(target_arch = "wasm32"))]
use crate::ocr::http_simple::HttpOcrEngine;
#[cfg(feature = "tesseract")]
use crate::ocr::tesseract::TesseractOcrEngine;
use crate::ocr_merge;
use crate::output::markdown;
use crate::projection;
#[cfg(not(target_arch = "wasm32"))]
use crate::render;
use crate::types::{ExtractedImage, OutlineTarget, Page, ParsedPage, PdfInput};
use pdfium::Library;

/// Result of parsing a document.
pub struct ParseResult {
    /// Parsed pages with projected text layout.
    pub pages: Vec<ParsedPage>,
    /// Full document text, concatenated from all pages.
    pub text: String,
    /// Document outline (bookmarks) when present. Used by the markdown
    /// emitter as a high-priority heading source on untagged PDFs.
    pub outline: Vec<OutlineTarget>,
    /// Raster images extracted from the document. Empty unless the parser
    /// was configured with `ImageMode::Embed`. Each entry carries the same
    /// `id` the markdown emitter referenced in `![](image_{id}.png)`, so the
    /// caller can match them up without parsing markdown.
    pub images: Vec<ExtractedImage>,
}

/// Result of rendering a single page screenshot.
#[derive(Debug, Clone)]
pub struct ScreenshotResult {
    pub page_num: u32,
    pub width: u32,
    pub height: u32,
    pub image_bytes: Vec<u8>,
}

/// Env var pointing at a fragmented glyph-outline → unicode font database
/// directory (`%02x%02x.msgpack` shards). When set, [`LiteParse::new`]
/// auto-wires a [`crate::FontDbResolver`] so buggy/obfuscated-font glyphs are
/// recovered without any extra wiring. Unset (default) leaves the hook dormant.
#[cfg(not(target_arch = "wasm32"))]
const FONT_DB_DIR_ENV: &str = "LITEPARSE_FONT_DB_DIR";

/// Build the default glyph resolver from the environment, if configured.
#[cfg(not(target_arch = "wasm32"))]
fn default_glyph_resolver() -> Option<std::sync::Arc<dyn crate::GlyphResolver>> {
    let dir = std::env::var_os(FONT_DB_DIR_ENV)?;
    if dir.is_empty() {
        return None;
    }
    Some(std::sync::Arc::new(crate::FontDbResolver::new(dir)))
}

#[cfg(target_arch = "wasm32")]
fn default_glyph_resolver() -> Option<std::sync::Arc<dyn crate::GlyphResolver>> {
    None
}

/// Main LiteParse orchestrator.
///
/// ### Thread safety
///
/// `LiteParse` is `Send + Sync` and safe to share across threads (e.g.
/// behind an `Arc`, or used concurrently from a multi-threaded `tokio`
/// runtime).
///
/// PDFium itself is **not** thread-safe, so all PDFium FFI work — document
/// loading, page rendering, text extraction — is serialized through a
/// process-global lock held by [`pdfium::Library`]. From a caller's
/// perspective, this means concurrent `parse_*` / `screenshot*` calls are
/// safe but their PDFium portions run sequentially. The OCR pass and grid
/// projection (which dominate runtime for OCR-heavy documents) run outside
/// the lock and remain fully concurrent.
pub struct LiteParse {
    config: LiteParseConfig,
    /// Optional caller-provided OCR engine. When set, this overrides the
    /// built-in selection logic (HTTP OCR / Tesseract). This is the primary
    /// mechanism for plugging an OCR engine in environments without the
    /// built-ins (e.g. WASM, where the JS side supplies a callback engine).
    ocr_engine_override: Option<std::sync::Arc<dyn OcrEngine>>,
    /// Optional caller-provided glyph recovery hook. When set, it is consulted
    /// as a last resort for buggy/obfuscated-font glyphs that liteparse's
    /// built-in cmap/AGL recovery could not decode. The published package ships
    /// none; the platform build injects an outline → unicode font-DB resolver.
    glyph_resolver: Option<std::sync::Arc<dyn crate::GlyphResolver>>,
}

impl LiteParse {
    pub fn new(config: LiteParseConfig) -> Self {
        Self {
            config,
            ocr_engine_override: None,
            glyph_resolver: default_glyph_resolver(),
        }
    }

    /// Override the OCR engine. When set, the engine is used regardless of
    /// `ocr_server_url` / built-in Tesseract availability.
    pub fn with_ocr_engine(mut self, engine: std::sync::Arc<dyn OcrEngine>) -> Self {
        self.ocr_engine_override = Some(engine);
        self
    }

    /// Inject a glyph recovery hook. When set, glyphs that liteparse considers
    /// untrusted and cannot decode with its built-in cmap/AGL recovery are
    /// passed to the resolver as vector-outline segments for a final attempt.
    pub fn with_glyph_resolver(
        mut self,
        resolver: std::sync::Arc<dyn crate::GlyphResolver>,
    ) -> Self {
        self.glyph_resolver = Some(resolver);
        self
    }

    /// Parse the configured `target_pages` string (e.g. `"1-5,10"`) into an
    /// explicit page list, or `None` when no selection was configured.
    fn resolve_target_pages(&self) -> Result<Option<Vec<u32>>, LiteParseError> {
        self.config
            .target_pages
            .as_ref()
            .map(|s| parse_target_pages(s))
            .transpose()
            .map_err(|e| format!("invalid --target-pages: {}", e).into())
    }

    /// Determine the complexity of each page in a document, returning a vector
    /// of `PageComplexityStats` for each page. This is useful for deciding
    /// whether to enable OCR on a per-page basis, or for other heuristics.
    pub async fn is_complex(
        &self,
        input: PdfInput,
    ) -> Result<Vec<ocr_merge::PageComplexityStats>, LiteParseError> {
        let log = |msg: &str| {
            if !self.config.quiet {
                eprintln!("{}", msg);
            }
        };

        let t0 = web_time::Instant::now();

        #[cfg(not(target_arch = "wasm32"))]
        let (validated_input, _guard) =
            conversion::resolve_pdf_input(input, self.config.password.as_deref(), false).await?;

        #[cfg(target_arch = "wasm32")]
        let validated_input = input;

        // Determine which pages to extract
        let target_pages = self.resolve_target_pages()?;

        // Load the document and extract text items. Complexity signals derive
        // from the text layer and page objects only — embedded image rasters
        // and hyperlinks are irrelevant here, so both are skipped to keep this
        // pass fast (its whole purpose is a cheap pre-OCR check).
        let password = self.config.password.as_deref();

        let lib = Library::init();
        let document = extract::load_document_from_input(&lib, &validated_input, password)?;

        let (pages, _) = extract::extract_pages_and_images(
            &document,
            target_pages.as_deref(),
            self.config.max_pages,
            false, // render_images: image rasters not needed for complexity
            false, // extract_links: irrelevant for complexity stats
            self.glyph_resolver.as_deref(),
            false, // emit_word_boxes: word boxes not needed for complexity stats
        )?;
        let t_extract = web_time::Instant::now();
        log(&format!(
            "[liteparse] extract: {:.1}ms ({} pages)",
            t_extract.duration_since(t0).as_secs_f64() * 1000.0,
            pages.len()
        ));

        let t_complexity = web_time::Instant::now();
        let page_complexities = pages
            .iter()
            .map(|page| {
                let page_obj = document.page((page.page_number - 1) as i32)?;
                ocr_merge::calculate_page_complexity(page, &page_obj)
            })
            .collect::<Result<Vec<_>, _>>()?;
        log(&format!(
            "[liteparse] complexity: {:.1}ms",
            t_complexity.duration_since(t_extract).as_secs_f64() * 1000.0
        ));

        Ok(page_complexities)
    }

    /// Parse a document from a file path, returning structured results.
    ///
    /// Non-PDF files are automatically converted to PDF first (requires
    /// LibreOffice/ImageMagick on the system).
    ///
    /// Not available on `wasm32` — the browser has no filesystem. Use
    /// [`LiteParse::parse_input`] with [`PdfInput::Bytes`] instead.
    #[cfg(not(target_arch = "wasm32"))]
    pub async fn parse(&self, input: &str) -> Result<ParseResult, LiteParseError> {
        self.parse_input(PdfInput::Path(input.to_string())).await
    }

    /// Parse a document from either a file path or raw bytes.
    ///
    /// Use `PdfInput::Path` for files on disk or `PdfInput::Bytes` for
    /// in-memory PDF data (e.g. from a network response or Node.js Buffer).
    pub async fn parse_input(&self, input: PdfInput) -> Result<ParseResult, LiteParseError> {
        let log = |msg: &str| {
            if !self.config.quiet {
                eprintln!("{}", msg);
            }
        };

        let t0 = web_time::Instant::now();

        #[cfg(not(target_arch = "wasm32"))]
        let (validated_input, _guard) =
            conversion::resolve_pdf_input(input, self.config.password.as_deref(), false).await?;

        #[cfg(target_arch = "wasm32")]
        let validated_input = input;

        // Determine which pages to extract
        let target_pages = self.resolve_target_pages()?;

        // Extract text (and pre-render OCR pages in one PDF load when OCR is on).
        // The PDFium lock is acquired for this entire critical section and
        // released before any `.await` below — OCR (network / CPU) and grid
        // projection (pure Rust) do not touch PDFium, so they can run
        // concurrently with other `LiteParse` calls.
        let password = self.config.password.as_deref();
        let render_images = matches!(self.config.image_mode, crate::config::ImageMode::Embed);

        // Build the OCR engine up front so the renderer knows whether to emit a
        // grayscale buffer (cheaper, for engines that binarize internally) or RGB.
        let ocr_engine: Option<std::sync::Arc<dyn OcrEngine>> = if self.config.ocr_enabled {
            Some(if let Some(e) = self.ocr_engine_override.clone() {
                e
            } else {
                #[cfg(not(target_arch = "wasm32"))]
                {
                    if let Some(ref url) = self.config.ocr_server_url {
                        std::sync::Arc::new(
                            HttpOcrEngine::with_headers(
                                url.clone(),
                                self.config.ocr_server_headers.clone(),
                            )
                            .with_retry(
                                crate::ocr::http_simple::OcrRetryConfig {
                                    hedge_delays_ms: self.config.ocr_hedge_delays_ms.clone(),
                                    ..Default::default()
                                },
                            ),
                        )
                    } else {
                        #[cfg(feature = "tesseract")]
                        {
                            std::sync::Arc::new(TesseractOcrEngine::new(
                                self.config.tessdata_path.clone(),
                            ))
                        }
                        #[cfg(not(feature = "tesseract"))]
                        {
                            return Err("OCR enabled but no --ocr-server-url provided and tesseract feature is disabled".into());
                        }
                    }
                }
                #[cfg(target_arch = "wasm32")]
                {
                    return Err(
                        "OCR enabled but no `ocrEngine` callback was provided (WASM builds have no built-in OCR engine)".into(),
                    );
                }
            })
        } else {
            None
        };
        let ocr_grayscale = ocr_engine.as_ref().is_some_and(|e| e.prefers_grayscale());

        let (pages, ocr_rendered, outline, images) = {
            let lib = Library::init();
            let document = extract::load_document_from_input(&lib, &validated_input, password)?;
            let outline = extract::extract_outline(&document);
            let (pages, images) = extract::extract_pages_and_images(
                &document,
                target_pages.as_deref(),
                self.config.max_pages,
                render_images,
                self.config.extract_links
                    && self.config.output_format == crate::config::OutputFormat::Markdown,
                self.glyph_resolver.as_deref(),
                self.config.emit_word_boxes,
            )?;
            let t_extract = web_time::Instant::now();
            log(&format!(
                "[liteparse] extract: {:.1}ms ({} pages)",
                t_extract.duration_since(t0).as_secs_f64() * 1000.0,
                pages.len()
            ));
            let rendered = if self.config.ocr_enabled {
                let r = ocr_merge::render_pages_for_ocr(
                    &document,
                    &pages,
                    self.config.dpi,
                    ocr_grayscale,
                )?;
                log(&format!(
                    "[liteparse] ocr render: {:.1}ms ({} pages)",
                    web_time::Instant::now()
                        .duration_since(t_extract)
                        .as_secs_f64()
                        * 1000.0,
                    r.len()
                ));
                r
            } else {
                Vec::new()
            };
            // `lib` is dropped here, releasing the PDFium lock.
            (pages, rendered, outline, images)
        };
        let mut pages = pages;
        let t1 = web_time::Instant::now();

        // OCR pass (engine resolved before the render block above).
        if let Some(engine) = ocr_engine {
            ocr_merge::ocr_and_merge_rendered(
                &mut pages,
                ocr_rendered,
                self.config.dpi,
                engine,
                &self.config.ocr_language,
                self.config.num_workers,
                self.config.ocr_failure_fatal,
            )
            .await?;
        }
        let t_ocr = web_time::Instant::now();
        log(&format!(
            "[liteparse] ocr: {:.1}ms",
            t_ocr.duration_since(t1).as_secs_f64() * 1000.0
        ));

        // Caller-requested content filters (page-region crop, diagonal-text
        // removal). Runs after OCR merge so it also drops OCR text outside the
        // crop region, and before projection so filtered items never surface.
        extract::apply_content_filters(
            &mut pages,
            self.config.crop_box.as_ref(),
            self.config.skip_diagonal_text,
        );

        // Grid projection
        let mut parsed_pages = projection::project_pages_to_grid(pages);
        let t2 = web_time::Instant::now();
        log(&format!(
            "[liteparse] project: {:.1}ms",
            t2.duration_since(t_ocr).as_secs_f64() * 1000.0
        ));

        let full_text = if self.config.output_format == crate::config::OutputFormat::Markdown {
            let page_md =
                markdown::format_markdown_pages(&parsed_pages, &outline, self.config.image_mode);
            let md = page_md.join("\n\n-----\n\n");
            for (page, md) in parsed_pages.iter_mut().zip(page_md) {
                page.markdown = md;
            }
            let t3 = web_time::Instant::now();
            log(&format!(
                "[liteparse] markdown: {:.1}ms",
                t3.duration_since(t2).as_secs_f64() * 1000.0
            ));
            md
        } else {
            parsed_pages
                .iter()
                .map(|p| p.text.as_str())
                .collect::<Vec<_>>()
                .join("\n\n")
        };

        let total = web_time::Instant::now().duration_since(t0).as_secs_f64() * 1000.0;
        log(&format!("[liteparse] total: {:.1}ms", total));

        Ok(ParseResult {
            pages: parsed_pages,
            text: full_text,
            outline,
            images,
        })
    }

    /// Parse from pre-extracted pages, skipping PDFium text extraction.
    ///
    /// The caller supplies `Page`s already populated with text items (and,
    /// optionally, graphics / struct nodes / image refs) in viewport space
    /// (top-left origin, 72 DPI). This runs only grid projection and the
    /// configured output formatter, so it touches neither PDFium nor OCR and
    /// is fully synchronous. Used when an external extractor (e.g. with its
    /// own font-recovery pipeline) owns text extraction.
    pub fn parse_from_pages(&self, pages: Vec<Page>, outline: Vec<OutlineTarget>) -> ParseResult {
        let mut parsed_pages = projection::project_pages_to_grid(pages);

        let full_text = if self.config.output_format == crate::config::OutputFormat::Markdown {
            let page_md =
                markdown::format_markdown_pages(&parsed_pages, &outline, self.config.image_mode);
            let md = page_md.join("\n\n-----\n\n");
            for (page, md) in parsed_pages.iter_mut().zip(page_md) {
                page.markdown = md;
            }
            md
        } else {
            parsed_pages
                .iter()
                .map(|p| p.text.as_str())
                .collect::<Vec<_>>()
                .join("\n\n")
        };

        ParseResult {
            pages: parsed_pages,
            text: full_text,
            outline,
            images: Vec::new(),
        }
    }

    /// Generate screenshots of document pages as PNG bytes.
    ///
    /// Non-PDF files are automatically converted to PDF first (requires
    /// LibreOffice/ImageMagick on the system). Plain-text formats cannot be
    /// rendered and return a clear error.
    #[cfg(not(target_arch = "wasm32"))]
    pub async fn screenshot(
        &self,
        input: &str,
        page_numbers: Option<Vec<u32>>,
    ) -> Result<Vec<ScreenshotResult>, LiteParseError> {
        self.screenshot_input(PdfInput::Path(input.to_string()), page_numbers)
            .await
    }

    /// Generate screenshots from a file path or raw bytes.
    #[cfg(not(target_arch = "wasm32"))]
    pub async fn screenshot_input(
        &self,
        input: PdfInput,
        page_numbers: Option<Vec<u32>>,
    ) -> Result<Vec<ScreenshotResult>, LiteParseError> {
        let log = |msg: &str| {
            if !self.config.quiet {
                eprintln!("{}", msg);
            }
        };

        let (validated_input, _guard) =
            conversion::resolve_pdf_input(input, self.config.password.as_deref(), true).await?;

        if let PdfInput::Path(ref path) = validated_input
            && !conversion::is_pdf(path)
        {
            log("[liteparse] converted input to PDF for screenshot rendering");
        }

        let rendered = render::render_pages_to_png(
            &validated_input,
            page_numbers.as_deref(),
            self.config.dpi,
            self.config.password.as_deref(),
        )?;

        Ok(rendered
            .into_iter()
            .map(|page| ScreenshotResult {
                page_num: page.page_num,
                width: page.width,
                height: page.height,
                image_bytes: page.png_bytes,
            })
            .collect())
    }

    pub fn config(&self) -> &LiteParseConfig {
        &self.config
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[allow(clippy::field_reassign_with_default)]
    fn test_new_stores_config() {
        let mut cfg = LiteParseConfig::default();
        cfg.ocr_enabled = false;
        cfg.max_pages = 7;
        let lp = LiteParse::new(cfg);
        assert!(!lp.config().ocr_enabled);
        assert_eq!(lp.config().max_pages, 7);
    }
}

File: packages/python/liteparse/parser.py
"""LiteParse Python wrapper - native Rust bindings via PyO3."""

from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple, Union

from liteparse._liteparse import LiteParse as _NativeLiteParse
from liteparse._liteparse import search_items as _native_search_items

from .types import (
    ExtractedImage,
    LiteParseConfig,
    PageComplexityStats,
    ParsedPage,
    ParseError,
    ParseResult,
    ScreenshotResult,
    TextItem,
    WordBox,
)


def _convert_native_result(native_result: Any) -> ParseResult:
    """Convert a native PyParseResult to our Python ParseResult."""
    pages: List[ParsedPage] = []
    for native_page in native_result.pages:
        text_items = [
            TextItem(
                text=item.text,
                x=item.x,
                y=item.y,
                width=item.width,
                height=item.height,
                font_name=item.font_name,
                font_size=item.font_size,
                confidence=item.confidence,
                rotation=getattr(item, "rotation", 0.0),
                words=[
                    WordBox(
                        text=w.text,
                        x=w.x,
                        y=w.y,
                        width=w.width,
                        height=w.height,
                    )
                    for w in getattr(item, "words", [])
                ],
            )
            for item in native_page.text_items
        ]
        pages.append(
            ParsedPage(
                page_num=native_page.page_num,
                width=native_page.width,
                height=native_page.height,
                text=native_page.text,
                markdown=native_page.markdown,
                text_items=text_items,
            )
        )
    images = [
        ExtractedImage(
            id=img.id,
            page=img.page,
            format=img.format,
            bytes=img.bytes,
        )
        for img in getattr(native_result, "images", [])
    ]
    return ParseResult(
        pages=pages,
        text=native_result.text,
        images=images,
    )


class LiteParse:
    """
    Python wrapper for the LiteParse document parser.

    Uses native Rust bindings for fast, in-process PDF parsing.

    Example:
        >>> from liteparse import LiteParse
        >>> parser = LiteParse()
        >>> result = parser.parse("document.pdf")
        >>> print(result.text)
    """

    def __init__(
        self,
        *,
        ocr_enabled: Optional[bool] = None,
        ocr_server_url: Optional[str] = None,
        ocr_server_headers: Optional[Dict[str, str]] = None,
        ocr_language: Optional[str] = None,
        tessdata_path: Optional[str] = None,
        max_pages: Optional[int] = None,
        target_pages: Optional[str] = None,
        dpi: Optional[float] = None,
        output_format: Optional[str] = None,
        preserve_very_small_text: Optional[bool] = None,
        password: Optional[str] = None,
        quiet: Optional[bool] = None,
        num_workers: Optional[int] = None,
        image_mode: Optional[str] = None,
        extract_links: Optional[bool] = None,
        ocr_failure_fatal: Optional[bool] = None,
        ocr_hedge_delays_ms: Optional[List[int]] = None,
        emit_word_boxes: Optional[bool] = None,
        crop_box: Optional[Tuple[float, float, float, float]] = None,
        skip_diagonal_text: Optional[bool] = None,
    ):
        """
        Initialize LiteParse parser.

        Args:
            ocr_enabled: Whether to enable OCR for scanned documents (default: True)
            ocr_server_url: URL of HTTP OCR server (uses Tesseract if not provided)
            ocr_server_headers: Extra HTTP headers sent with every request to
                ``ocr_server_url`` (e.g. ``{"Authorization": "Bearer <token>"}``)
            ocr_language: Language code for OCR (e.g., "eng", "fra")
            tessdata_path: Path to tessdata directory for Tesseract
            max_pages: Maximum number of pages to parse
            target_pages: Specific pages to parse (e.g., "1-5,10,15-20")
            dpi: DPI for rendering (affects OCR quality)
            output_format: Output format: "json", "text", or "markdown" (default: "json")
            preserve_very_small_text: Whether to preserve very small text
            password: Password for encrypted/protected documents
            quiet: Suppress progress output
            num_workers: Number of concurrent OCR workers (default: CPU cores - 1)
            extract_links: Render hyperlink annotations as ``[text](url)`` in
                markdown output (default: True). Set False for plain anchor text.
            ocr_failure_fatal: Whether a systemic OCR failure (every OCR task
                failed and at least one was a text-sparse page) aborts the whole
                parse (default: True). Set False to keep already-recovered native
                text and return partial results instead of raising — for callers
                that prefer a degraded document over a hard failure.
            ocr_hedge_delays_ms: Request-hedging schedule for HTTP OCR, in
                milliseconds. Empty or single-element means no hedging (one
                request per attempt — the default). With multiple delays (e.g.
                ``[0, 5000, 10000]``) each attempt fires a duplicate request at
                every delay and takes the first to succeed, cancelling the rest
                — trades extra OCR-server load for lower tail latency.
            emit_word_boxes: Emit per-word sub-boxes on each text item
                (``TextItem.words``). Default False. Word boxes roughly double
                the text-item payload, so enable only for word-level bbox
                attribution.
            crop_box: Restrict output to a page sub-region, as a
                ``(top, right, bottom, left)`` tuple where each value is the
                fraction cropped from that side (top-left origin, each in
                ``[0, 1]``); e.g. ``(0, 0, 0, 0.5)`` keeps the right half. A
                text item survives only if it lies entirely inside the
                remaining rectangle. None (default) keeps the whole page.
            skip_diagonal_text: Drop diagonal text — items whose rotation is
                more than 2° off the nearest right angle (0/90/180/270).
                Default False. Use to exclude rotated watermarks/stamps.
        """
        kwargs = {}
        if ocr_enabled is not None:
            kwargs["ocr_enabled"] = ocr_enabled
        if ocr_server_url is not None:
            kwargs["ocr_server_url"] = ocr_server_url
        if ocr_server_headers is not None:
            kwargs["ocr_server_headers"] = ocr_server_headers
        if ocr_language is not None:
            kwargs["ocr_language"] = ocr_language
        if tessdata_path is not None:
            kwargs["tessdata_path"] = tessdata_path
        if max_pages is not None:
            kwargs["max_pages"] = max_pages
        if target_pages is not None:
            kwargs["target_pages"] = target_pages
        if dpi is not None:
            kwargs["dpi"] = dpi
        if output_format is not None:
            kwargs["output_format"] = output_format
        if preserve_very_small_text is not None:
            kwargs["preserve_very_small_text"] = preserve_very_small_text
        if password is not None:
            kwargs["password"] = password
        if quiet is not None:
            kwargs["quiet"] = quiet
        if num_workers is not None:
            kwargs["num_workers"] = num_workers
        if image_mode is not None:
            kwargs["image_mode"] = image_mode
        if extract_links is not None:
            kwargs["extract_links"] = extract_links
        if ocr_failure_fatal is not None:
            kwargs["ocr_failure_fatal"] = ocr_failure_fatal
        if ocr_hedge_delays_ms is not None:
            kwargs["ocr_hedge_delays_ms"] = ocr_hedge_delays_ms
        if emit_word_boxes is not None:
            kwargs["emit_word_boxes"] = emit_word_boxes
        if crop_box is not None:
            kwargs["crop_box"] = crop_box
        if skip_diagonal_text is not None:
            kwargs["skip_diagonal_text"] = skip_diagonal_text

        self._native = _NativeLiteParse(**kwargs)

    def parse(
        self,
        file_data: Union[str, Path, bytes],
    ) -> ParseResult:
        """
        Parse a document file.

        Args:
            file_data: Path to the document file, or raw PDF bytes.

        Returns:
            ParseResult containing the parsed document data.

        Raises:
            ParseError: If parsing fails.
            FileNotFoundError: If the file doesn't exist.
        """
        try:
            if isinstance(file_data, bytes):
                native_result = self._native.parse_bytes(file_data)
            else:
                file_path = Path(file_data)
                if not file_path.exists():
                    raise FileNotFoundError(f"File not found: {file_path}")
                native_result = self._native.parse(str(file_path.absolute()))
            return _convert_native_result(native_result)
        except FileNotFoundError:
            raise
        except Exception as e:
            raise ParseError(str(e)) from e

    def is_complex(
        self,
        file_data: Union[str, Path, bytes],
    ) -> List[PageComplexityStats]:
        """
        Determine per-page complexity without running a full parse.

        Returns one entry per page with signals (text coverage, images, garbled
        text, vector area) and a ``needs_ocr`` verdict — a cheap pre-OCR check to
        decide whether a document needs advanced parsing.

        Args:
            file_data: Path to the document file, or raw PDF bytes.

        Returns:
            List of PageComplexityStats, one per page.

        Raises:
            ParseError: If the check fails.
            FileNotFoundError: If the file doesn't exist.
        """
        try:
            if isinstance(file_data, bytes):
                native_stats = self._native.is_complex_bytes(file_data)
            else:
                file_path = Path(file_data)
                if not file_path.exists():
                    raise FileNotFoundError(f"File not found: {file_path}")
                native_stats = self._native.is_complex(str(file_path.absolute()))
            return [
                PageComplexityStats(
                    page_number=s.page_number,
                    text_length=s.text_length,
                    text_coverage=s.text_coverage,
                    has_substantial_images=s.has_substantial_images,
                    image_block_count=s.image_block_count,
                    image_coverage=s.image_coverage,
                    largest_image_coverage=s.largest_image_coverage,
                    full_page_image=s.full_page_image,
                    uncovered_vector_area=s.uncovered_vector_area,
                    is_garbled=s.is_garbled,
                    page_area=s.page_area,
                    needs_ocr=s.needs_ocr,
                    reasons=list(s.reasons),
                )
                for s in native_stats
            ]
        except FileNotFoundError:
            raise
        except Exception as e:
            raise ParseError(str(e)) from e

    def screenshot(
        self,
        file_path: Union[str, Path],
        *,
        page_numbers: Optional[List[int]] = None,
    ) -> List[ScreenshotResult]:
        """
        Generate screenshots of document pages.

        Supports PDFs natively. Non-PDF formats (DOCX, XLSX, images, etc.) are
        automatically converted to PDF before rendering when the required system
        tools are installed. Plain-text formats cannot be rendered.

        Args:
            file_path: Path to the document file (PDF, DOCX, images, etc.).
            page_numbers: Specific page numbers to screenshot (1-indexed).
                          If None, screenshots all pages.

        Returns:
            List of ScreenshotResult with PNG image bytes.

        Raises:
            FileNotFoundError: If the file doesn't exist.
            ParseError: If screenshot generation fails.
        """
        file_path = Path(file_path)
        if not file_path.exists():
            raise FileNotFoundError(f"File not found: {file_path}")

        try:
            native_results = self._native.screenshot(
                str(file_path.absolute()),
                page_numbers,
            )
            return [
                ScreenshotResult(
                    page_num=r.page_num,
                    width=r.width,
                    height=r.height,
                    image_bytes=r.image_bytes,
                )
                for r in native_results
            ]
        except Exception as e:
            raise ParseError(str(e)) from e

    def get_config(self) -> LiteParseConfig:
        """Return the resolved configuration."""
        cfg = self._native.config
        return LiteParseConfig(
            ocr_language=cfg.ocr_language,
            ocr_enabled=cfg.ocr_enabled,
            ocr_server_url=cfg.ocr_server_url,
            ocr_server_headers=cfg.ocr_server_headers,
            tessdata_path=cfg.tessdata_path,
            max_pages=cfg.max_pages,
            target_pages=cfg.target_pages,
            dpi=cfg.dpi,
            output_format=cfg.output_format,
            preserve_very_small_text=cfg.preserve_very_small_text,
            password=cfg.password,
            quiet=cfg.quiet,
            num_workers=cfg.num_workers,
        )

    def __repr__(self) -> str:
        return "LiteParse()"


def search_items(
    items: List[TextItem],
    phrase: str,
    *,
    case_sensitive: bool = False,
) -> List[TextItem]:
    """
    Search text items for phrase matches, returning merged items with combined bounding boxes.

    A phrase may span multiple adjacent text items. This function concatenates
    consecutive items, finds matches, and returns synthetic merged TextItem
    objects with combined bounding boxes.

    Args:
        items: List of TextItem objects to search through.
        phrase: The phrase to search for.
        case_sensitive: Whether the search should be case-sensitive (default: False).

    Returns:
        List of TextItem objects representing the matched regions.
    """
    native_results = _native_search_items(items, phrase, case_sensitive=case_sensitive)
    return [
        TextItem(
            text=item.text,
            x=item.x,
            y=item.y,
            width=item.width,
            height=item.height,
            font_name=item.font_name,
            font_size=item.font_size,
            confidence=item.confidence,
        )
        for item in native_results
    ]
